tensorflow/core/kernels/conv_ops_gpu_3.cu.cc

```diff
@@ -413,6 +429,275 @@ struct PadInput<GPUDevice, T, int, NDIMS> {
+                                           int total_tiles_count,
+                                           const T* input,
+                                           const Dimension<3>& input_dims,
+                                           T* output) {
+
+    static_assert((TileLongSide & (TileLongSide - 1)) == 0,
+      "The length of the longer side of the tile is always a power of 2.");
+
+    const int NumThread = TileLongSide;
+    if (tile_size_i <= TileLongSide && tile_size_j <= TileShortSide)
+      SwapDimension1And2InTensor3UsingTiles<
+          T, NumThread, TileLongSide,
+          TileShortSide><<<total_tiles_count, NumThread, 0, d.stream()>>>(
+          input, input_dims, output);
+    else if (tile_size_j <= TileLongSide && tile_size_i <= TileShortSide)
+      SwapDimension1And2InTensor3UsingTiles<
+          T, NumThread, TileShortSide,
+          TileLongSide><<<total_tiles_count, NumThread, 0, d.stream()>>>(
+          input, input_dims, output);
+    return;
+  }
+};
+
+template <typename T, int TileLongSide, int TileShortSide>
+struct BatchNarrowMatrixTransposeDispatcher<
+    T, TileLongSide, TileShortSide,
+    typename std::enable_if<TileSizePossibilityFrontierCheck(
+                                TileLongSide, TileShortSide, sizeof(T),
+                                /* mode = outside */ 1),
+                            void>::type> {
+  static void DoIt(const GPUDevice& d, int tile_size_i,
+                                           int tile_size_j,
+                                           int total_tiles_count,
+                                           const T* input,
+                                           const Dimension<3>& input_dims,
+                                           T* output) {
+    assert(false &&
+           "BatchNarrowMatrixTransposeDispatcher has requested an unexpected "
+           "launch configuration. ");
+  }
+};
+
+// A helper function to make RunSwapDimension1And2InTensor3 concise. This
+// helper function looks at the data type and input matrix sizes and decides
+// the thread numbers and tile sizes to use.
+template <typename T>
+void SwapDimension1And2InTensor3WithNarrowMatrices(
+    const GPUDevice& d, const T* input, const Dimension<3>& input_dims,
+    T* output, const int kMinDimensionToUseTiles) {
+  // Define available tile sizes here for each size of data type supported:
+  std::vector<std::pair<int, int>> tile_spec =
+      []() -> std::vector<std::pair<int, int>> {
+    switch (sizeof(T)) {
+      case 1:
+        return {{32, 15}, {64, 15}, {128, 15}, {256, 15}, {512, 4}, {1024, 2}};
+      case 2:
+        return {{32, 15}, {64, 15}, {128, 15}, {256, 15}, {512, 4}, {1024, 2}};
+      case 4:
+        return {{32, 15}, {64, 15}, {128, 15}, {256, 15}, {512, 4}, {1024, 2}};
+      case 8:
+        return {{32, 15}, {64, 15}, {128, 8}, {256, 4}, {512, 2}};
+      case 16:
+        return {{32, 4},  {64, 4},  {128, 4}, {256, 2}};
+    }
+    return {};
```

Comment:
Nit: I think this would be clearer as a `default`.  But also...should we have some static assert that checks that `sizeof(T) <= 16`?